Rebus.AmazonS3/AmazonS3/AmazonS3SubscriptionsStorage.cs

```diff
@@ -0,0 +1,230 @@
+        private readonly Func<IAmazonS3> _amazonS3Factory;
+        private readonly ILog _log;
+
+        public AmazonS3SubscriptionsStorage(AWSCredentials credentials, AmazonS3Config amazonS3Config, AmazonS3DataBusOptions options, IRebusLoggerFactory rebusLoggerFactory)
+            : this(rebusLoggerFactory)
+        {
+            _awsCredentials = credentials ?? throw new ArgumentNullException(nameof(credentials));
+            _awsConfig = amazonS3Config ?? throw new ArgumentNullException(nameof(amazonS3Config));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        public AmazonS3SubscriptionsStorage(Func<IAmazonS3> amazonS3Factory, IRebusLoggerFactory rebusLoggerFactory)
+            : this(rebusLoggerFactory)
+        {
+            _amazonS3Factory = amazonS3Factory;
+        }
+
+        private AmazonS3SubscriptionsStorage(IRebusLoggerFactory rebusLoggerFactory)
+        {
+            _log = rebusLoggerFactory?.GetLogger<AmazonS3SubscriptionsStorage>() ?? throw new ArgumentNullException(nameof(rebusLoggerFactory));
+        }
+
+        public bool IsCentralized
+        {
+            get { return true; }
+        }
+
+        public async Task<string[]> GetSubscriberAddresses(string topic)
+        {
+            await EnsureBucketExist();
+
+            var subscriptions = await GetSubscriptions(topic);
+
+            HashSet<string> subscribers;
+            if(subscriptions.TryGetValue(topic, out subscribers))
+                return subscribers.ToArray();
+
+            return new string[0];
+        }
+
+        public async Task RegisterSubscriber(string topic, string subscriberAddress)
+        {
+            await EnsureBucketExist();
+
+            var subscriptions = await GetSubscriptions(topic);
+
+            subscriptions
+                .GetOrAdd(topic, () => new HashSet<string>())
+                .Add(subscriberAddress);
+
+            await SaveToS3(subscriptions, topic);
+        }
+
+        public async Task UnregisterSubscriber(string topic, string subscriberAddress)
+        {
+            await EnsureBucketExist();
+
+            var subscriptions = await GetSubscriptions(topic);
+
+            subscriptions
+                .GetOrAdd(topic, () => new HashSet<string>())
+                .Remove(subscriberAddress);
+
+            await SaveToS3(subscriptions, topic);
+        }
+
+        private async Task<Dictionary<string, HashSet<string>>> GetSubscriptions(string topic)
+        {
+            if (await ObjectExists(topic))
+            {
+                string jsonText = await FetchFromS3(topic);
+
+                var subscriptions = JsonConvert.DeserializeObject<Dictionary<string, HashSet<string>>>(jsonText);
+
+                return subscriptions;
+            }
+
+            return new Dictionary<string, HashSet<string>>();
+        }
+
+        private async Task<bool> ObjectExists(string key)
```

Comment:
It is good practice to postfix any asynchronous method with `Async`, i.e. method name should be `ObjectExistsAsync`